04_Scheduling/03_Tasks_And_Threads.md

```diff
@@ -62,13 +62,109 @@ typedef struct
 
-```
+```c
 size_t next_free_pid = 0;
 ```  
 
+
 ### Virtual memory space
 
-TBD 
+On of the most useful features of having the paging enabled is that we have the concept of Virtual Memory available (for a complete definition refer to the [Paging](,,/02_Memory_Management/Paging.md) and [Virtual Memory](../02_Memory_Management/04_Virtual_Memory_Manager.md) chapter), and from a process point of view it means that we can have each process with it's own addressing space. 
+
+What are the improvements that we can have from adding this feature: 
+
+* The most important one is that it provides isolation between processes, so same address location in two different process will point to two different physicall addresses (unless we we want it otherwise in some special cases). 
+* Every process will have it's own memory heap (technically we can implement also different type of processes to use different type of memory allocators. 
+
+What we need to add this feature to our proceess? The paging in x86_64 bits is handled by a set of hierarchical page directories/tables, and the address is basically containing the entry number in those tables. The root address of this table is stored in the PDBR (if you have not clear how it works, go back and read the paging chapter), then the first thing we need to change in our process structure is to add a new variable to store the value of the PDBR register: 
+
+```c
+typedef struct {
+    size_t pid;
+    char name[NAME_MAX_LEN];
+    status_t process_status;
+    cpu_status_t context;
+    uint64_t pdbr;
+} process_t;
+
+```
+
+Adding the pdbr reference is not enough. Let's see why:
+
+* The first problem is that the PDBR contains the pointer to the root page directory of the virtual memory space (remember that an address with pagin enabled is a composition of table indexes), so this means we need to allocate for it.
+* A memory address to be accessibile, has to have the present flag set, and various directories/table entries to be properly configured
+* The step above means that when we create a new virtual memory space, we lose any reference to the kernel, so this means that as soon as we switch to the new pdbr, if the kernel is not mapped where it is suppose to be we will cause our kernel to crash, so we need to remap the kernel in the new process memory environment. How to map it is explained in the Paging chapter.
+
+Implementing them is just a variation of what we have seen in the Paging chapter, so we'll leave it as an exercies.
```

Comment:
Nice summary! Although this last line says the same thing as the previous line, it can probably be omitted.
There's also a few sneaky issues that can arise from having a separate page table per process later on. For example, if the kernel in one process maps something in the higher half, it will only appear in that process. If the kernel of another process (it's the same kernel, but with a different mapping now - because its a different process) tries to access a pointer in that region it will page fault. There's a bunch of ways to solve this, and it could probably get it's own section. Maybe I can write something on what I've tried so far.